includes/embeds/class-amp-youtube-embed-handler.php

```diff
@@ -82,117 +94,313 @@ public function unregister_embed() {
 	 *
 	 * @param string $cache Cache for oEmbed.
 	 * @param string $url   Embed URL.
+	 *
 	 * @return string Embed.
 	 */
 	public function filter_embed_oembed_html( $cache, $url ) {
-		$id = $this->get_video_id_from_url( $url );
-		if ( ! $id ) {
-			return $cache;
+
+		if ( empty( $cache ) || empty( $url ) ) {
+			return '';
 		}
 
-		$props = $this->parse_props( $cache, $url, $id );
-		if ( empty( $props ) ) {
+		$video_id = $this->get_video_id_from_url( $url );
+
+		if ( ! $video_id ) {
 			return $cache;
 		}
 
-		$props['video_id'] = $id;
-		return $this->render( $props, $url );
+		return $this->render( $cache, $url );
 	}
 
 	/**
-	 * Parse AMP component from iframe.
+	 * Convert YouTube iframe into AMP YouTube component.
 	 *
-	 * @param string $html     HTML.
-	 * @param string $url      Embed URL, for fallback purposes.
-	 * @param string $video_id YouTube video ID.
-	 * @return array|null Props for rendering the component, or null if unable to parse.
+	 * @param string $html HTML markup of YouTube iframe.
+	 * @param string $url  YouTube URL.
+	 *
+	 * @return string HTML markup of AMP YouTube component.
 	 */
-	private function parse_props( $html, $url, $video_id ) {
-		$props = $this->match_element_attributes( $html, 'iframe', [ 'title', 'height', 'width' ] );
-		if ( ! isset( $props ) ) {
-			return null;
+	public function render( $html, $url ) {
+
+		$attributes = $this->prepare_attributes( $url );
```

Comment:
All methods that call the `render()` and `prepare_attributes()` methods have already determined the video ID, so I'd suggest passing it to them so as not to re-parse it from the URL each time. Something like:

```diff
--- a/includes/embeds/class-amp-youtube-embed-handler.php	(revision db1bd35807612e6a89beba31fa7d05a2be5c049d)
+++ b/includes/embeds/class-amp-youtube-embed-handler.php	(date 1625728551158)
@@ -109,20 +109,21 @@
 			return $cache;
 		}
 
-		return $this->render( $cache, $url );
+		return $this->render( $cache, $url, $video_id );
 	}
 
 	/**
 	 * Convert YouTube iframe into AMP YouTube component.
 	 *
-	 * @param string $html HTML markup of YouTube iframe.
-	 * @param string $url  YouTube URL.
+	 * @param string $html      HTML markup of YouTube iframe.
+	 * @param string $url       YouTube URL.
+	 * @param string $video_id  YouTube video ID.
 	 *
 	 * @return string HTML markup of AMP YouTube component.
 	 */
-	public function render( $html, $url ) {
+	public function render( $html, $url, $video_id ) {
 
-		$attributes = $this->prepare_attributes( $url );
+		$attributes = $this->prepare_attributes( $url, $video_id );
 
 		$iframe_props = [ 'title', 'height', 'width' ];
 		$props        = $this->match_element_attributes( $html, 'iframe', $iframe_props );
@@ -203,7 +204,7 @@
 		}
 
 		$url        = $node->getAttribute( 'src' );
-		$attributes = $this->prepare_attributes( $url );
+		$attributes = $this->prepare_attributes( $url, $video_id );
 
 		if ( ! empty( $node->getAttribute( 'title' ) ) ) {
 			$attributes['title'] = $node->getAttribute( 'title' );
@@ -253,17 +254,12 @@
 	/**
 	 * Prepare attributes for amp-youtube component.
 	 *
-	 * @param string $url YouTube video URL.
+	 * @param string $url      YouTube video URL.
+	 * @param string $video_id YouTube video ID.
 	 *
 	 * @return array prepared arguments for amp-youtube component.
 	 */
-	private function prepare_attributes( $url ) {
-
-		$video_id = $this->get_video_id_from_url( $url );
-
-		if ( ! $video_id ) {
-			return [];
-		}
+	private function prepare_attributes( $url, $video_id ) {
 
 		$attributes = [
 			'data-videoid' => $video_id,
@@ -510,6 +506,6 @@
 			return $html;
 		}
 
-		return $this->render( $html, $attr['src'] );
+		return $this->render( $html, $attr['src'], $video_id );
 	}
 }
```